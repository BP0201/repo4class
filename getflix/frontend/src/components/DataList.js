import React, { useState, useEffect } from 'react';

import MovieAPI from '../apis/MovieApi';
import SearchForm from '../forms/SearchForm'
import DataCard from './DataCard';

/** DataList
 * Renders the SearchForm and an array of DataCards, displayed after a search.
 */

const DataList = ({ category = "multi", add, remove }) => {
    // console.debug("DataList");
    const [data, setData] = useState([]);
    // const [isFavorited, setIsFavorited] = useState(false);

    useEffect(function getDataOnMount() {
        // console.debug("DataList useEffect getDataOnMount")
            search();
    }, []);

    async function search(q) {
        const data = await MovieAPI.fetch(q, category)
        if (!(data.length && data[0].id === 954884)) {
            setData(data);
        }
    }

    function filter(category) {
        const filtered = data.filter(d => d.media_type === category);
        setData(filtered);
    }

    function sortByRelease(type) {
        let newData = [...data];
        function sortOld(a, b) {
            if (
            (a.release_date || a.first_air_date)
            <
            (b.release_date || b.first_air_date)
            ){
              return -1;
            }
            if (
            (a.release_date || a.first_air_date)
            >
            (b.release_date || b.first_air_date) ){
              return 1;
            }
            return 0;
          }
          function sortNew(a, b) {
            if (
            (a.release_date || a.first_air_date)
            <
            (b.release_date || b.first_air_date)
            ){
              return 1;
            }
            if (
            (a.release_date || a.first_air_date)
            >
            (b.release_date || b.first_air_date) ){
              return -1;
            }
            return 0;
          }
        if (type === "old") {
            newData.sort( sortOld );
        } else if (type === "new") {
            newData.sort( sortNew );
        }
        setData(newData);
    }

    if(!data) return <h1>Loading...</h1>

    function handleResults(data) {
        return data.map(d => (
            <DataCard
                name={d.title || d.name}
                key={d.id}
                id={d.id}
                poster={ d.poster_path }
                release_date={d.release_date || d.first_air_date}
                // genres={d.genres}
                description={d.overview}
                category={d.media_type}
                knownFor={d.known_for}
                handleResults2={handleResults2}
                add={add}
                remove={remove}
            />
        ))
    }

    function handleResults2(arr) {
        return arr.map(d => (
            <DataCard
                name={d.title || d.name}
                key={d.id}
                id={d.id}
                poster={ d.poster_path }
                release_date={d.release_date || d.first_air_date}
                // genres={d.genres}
                description={d.overview}
                category={d.media_type}
                knownFor={d.known_for}
                add={add}
                remove={remove}
            />
        ))
    }

    return (
        <div className='DataList container'>
            <h3 className='display-4 my-2'>Search</h3>
            {data.length ?
            <div className='button-group'>
            <button onClick={() => filter("tv")}>Hide Movies</button>
            <button onClick={() => filter("movie")}>Hide Shows</button>
            <button onClick={() => sortByRelease("old")}>Sort results by oldest</button>
            <button onClick={() => sortByRelease("new")}>Sort results by newest</button>
            </div>
            :
            ""
            }
            <SearchForm searchFor={search} />
            {data.length
            ? (
                <div className="DataList-list">
                  {handleResults(data)}
                </div>
            ) : (
                <p className="lead">Results will appear here.</p>
            )}
        </div>
    )
}

export default DataList;