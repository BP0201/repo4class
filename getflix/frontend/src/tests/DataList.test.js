import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import DataList from '../components/DataList';

describe("DataList", () => {
  it('matches snapshot', () => {
    const { asFragment, getByTestId } = render(<DataList />)
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("search-form")).toBeInTheDocument();
  });
});
