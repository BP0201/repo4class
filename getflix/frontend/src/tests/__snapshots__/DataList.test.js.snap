// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataList matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="DataList container"
  >
    <h3
      class="display-4 my-2"
    >
      Search
    </h3>
    <div
      class="SearchForm mb-4"
    >
      <form
        class="form-inline"
        data-testid="search-form"
      >
        <div
          class="input-group col-12"
        >
          <input
            class="form-control flex-grow-1"
            data-testid="search-form-input"
            name="searchTerm"
            placeholder="Type here"
            type="text"
            value=""
          />
          <div
            class="input-group-append"
          >
            <button
              class="btn btn-outline-info"
              data-testid="search-form-btn"
              type="submit"
            >
              Search
            </button>
          </div>
        </div>
      </form>
    </div>
    <p
      class="lead"
    >
      Results will appear here.
    </p>
  </div>
</DocumentFragment>
`;
